### PR TITLE
Migration to Keycloak 20.0.0 (Quarkus-based)

### PR DESCRIPTION
**Pull request by michaelkain**
- Migration of Keycloak to 20.0.0
- Update of RabbitMQ from 3.7.7 to 3.10.7

### CODE REVIEW

**Comment by michaelkain** [comment]
see https://github.com/fli-iam/shanoir-ng/issues/1346

**Comment by michaelkain** [comment]
@todo: migrate keycloak-core as well to container, and test

**Comment by michaelkain** [comment]
@jcomedouteau Hi Jean-Come! I have finished my local dev tests and the PR is ready for test from my side.
My tests:
- login page works, all links open
- could login with user admin
- all microservices work
- import + ohif viewer work
- bootstrap.sh runs with --clean
Please let me know if locally all works for you as well? Thank you!

**Comment by michaelkain** [comment]
@a-ba Hi Anthony! From my point of view from the pure dev side everything is finished for the migration.
for production you will have to enable mode "production" in the Dockerfile and test in prod what that implies.
Looking forward to see you back for the next weekly to exchange and with kind regards, Michael

**Comment by jcomedouteau** [comment]
Well done ! 
I'll check this on monday :)
Have a nice WE !
Jean-Côme

**Comment by jcomedouteau** [comment]
I have this error, I think we may need to update these files permissions:
sed: couldn't open temporary file /opt/keycloak/themes/shanoir-theme/login/sedrjhmJP: Permission denied

**Comment by michaelkain** [comment]
Hi Jean-Come, indeed your error is weird, and hard to fix for me as I do not have it. I pushed some lines, that might be worth a try on your machine. Have you maybe tried a docker system prune? I am wondering, as the only two lines of code, that I can relate with your error are in entrypoint, lines 26+27, and I have normally not touched much on that. Happy testing and with kind regards, Michael

**Comment by jcomedouteau** [comment]
Hi Michael,
I'll try locally with your changes. If it still does not work, I'll try the docker prune to fully reset all my docker.
Jean-Côme

**Comment by jcomedouteau** [comment]
It did not work, even with a docker prune. It may be a local problem, can you try to deploy on qualif maybe ?

**Comment by michaelkain** [comment]
THX Jean-Come! @pierrehenri-dauvergne Hi Pierre-Henri, could you please test on your machine, if you have the same problem under Windows as Jean-Come please? I can test on the qualif as well.

**Comment by pierrehenri-dauvergne** [comment]
Hello @michaelkain, everything is working fine on my side

**Comment by michaelkain** [comment]
Hi @pierrehenri-dauvergne thank you Pierre-Henri! I upgrade the KC lib versions too and will then test on qualif for the issue of Jean-Come. Cheers, Michael

**Comment by michaelkain** [comment]
I updated all depending libs to version 20.0.0 to harmonize everything: 1) shanoir-uploader 2) shanoir-ng-keycloak-auth 3) shanoir-ng-back (concerns all microservices). I tested on my machine: login works, all microservices work and last login date is updated and users are created in Keycloak.

**Comment by michaelkain** [comment]
I deploy on qualif to check

**Comment by michaelkain** [comment]
same issue on qualif, even when hard to see:
dk logs qualif-keycloak
sed: couldn't open temporary file /opt/keycloak/themes/shanoir-theme/login/sed63BJIC: Permission denied

**Comment by jcomedouteau** [comment]
Do you want to try to check this problem as I can test locally ?

**Comment by michaelkain** [comment]
Hi, I will try to figure out and do some tests on qualif, if it becomes to time demanding I would like to do a session on your machine with pleasure. Thank you!

**Comment by michaelkain** [comment]
Hi Jean-Come, @jcomedouteau , please test again, I think I passed by the problem. It goes further on qualif. I would say the error on qualif is now depending on qualif, where only Anthony can go further. Thx, Michael

**Comment by michaelkain** [comment]
@jcomedouteau Hi Jean-Come, it works on qualif, at least the permission error is solved. Could you please verify on your machine if it might finish? Thank you and have a good start into your week, Michael

**Comment by jcomedouteau** [comment]
Hi Michael, I'm on it !

**Review by jcomedouteau — APPROVED** [review]
It works!

**Comment by michaelkain** [comment]
Perfect, thank you Jean-Come! @pierrehenri-dauvergne Could you please try on Windows as well?

**Comment by michaelkain** [comment]
@a-ba Hi Anthony, I hope you are doing fine. I have the below error on qualif:
ln: target '/tmp/import/' is not a directory: No such file or directory
error: import dir '/import' does not contain any users file
aborting because of errors
Could you please have a look on that to progress with the Keycloak migration on prod?
Thank you and with kind regards, Michael

**Comment by pierrehenri-dauvergne** [comment]
> Perfect, thank you Jean-Come! @pierrehenri-dauvergne Could you please try on Windows as well?

I just tried and got no issue, everything seems fine on my side

**Comment by a-ba** [comment]
I have just run the bootstrap script but the admin console at http://localhost:8080/auth/admin/ won't load (it is frozen with the message  « Loading the admin console »).

Is the admin console working on your machines?

(note: i will have several fixes to submit, do not merge the PR now)

File: docker-compose/keycloak/Dockerfile [hunk 1]
```diff
@@ -10,19 +10,39 @@
 # You should have received a copy of the GNU General Public License
 # along with this program. If not, see https://www.gnu.org/licenses/gpl-3.0.html
 
-FROM jboss/keycloak:16.1.1
+#
+# Use builder to integrate custom provider
+#
+FROM quay.io/keycloak/keycloak:20.0.0 as builder
```
Comment (a-ba) [inline docker-compose/keycloak/Dockerfile:16]: They published 3 bugfix releases since 20.0.0 (current is 20.0.3)
https://quay.io/repository/keycloak/keycloak?tab=tags

we should rather use  `quay.io/keycloak/keycloak:20.0` as base image

File: docker-compose.yml [hunk 2]
```diff
@@ -50,7 +55,7 @@ services:
   #  
   rabbitmq:
     container_name: "${SHANOIR_PREFIX}rabbitmq"
-    image: rabbitmq:3.7.7-management-alpine
```
Comment (a-ba) [inline docker-compose.yml:53]: this change seems unrelated to keycloak

File: docker-compose/keycloak/Dockerfile [hunk 3]
```diff
@@ -10,19 +10,39 @@
...
+COPY shanoir-ng-keycloak-auth.jar /opt/keycloak/providers
 
-COPY shanoir-ng-keycloak-auth.jar /opt/jboss/keycloak/standalone/deployments/
+# Configure a database vendor
+ENV KC_DB=mysql
 
-COPY cfg/. /opt/jboss/keycloak/
+# Enable health and metrics support
+ENV KC_HEALTH_ENABLED=true
+ENV KC_METRICS_ENABLED=true
```
Comment (a-ba) [inline docker-compose/keycloak/Dockerfile:25]: Are the `/health` and `/metrics` endpoints actually used by shanoir ? If not i would rather let them disabled and avoid exposing any information (if any sysadmin wants them, he can set the env var and put an adequate filter in his reverse proxy so as not to expose them to the public).

**Comment by a-ba** [comment]
I pushed a series of fixes at https://github.com/michaelkain/shanoir-ng/pull/9

**Comment by a-ba** [comment]
https://github.com/michaelkain/shanoir-ng/pull/9 successfully tested in qualif

File: docker-compose.yml (hunk 2 above)
Comment (michaelkain) [reply docker-compose.yml:53]: Hi Anthony, I agree for the general point of view, but Shanoir does not work on my machine without that update and as it works as well on the other machines, I am happy to bring it into develop anyway. I will add it into the description, that we are not surprised if we might have side effects. As I tested imports on my machine, there should not be one except updating rabbitmq and using a newer version.

File: docker-compose/keycloak/Dockerfile (hunk 1 above)
Comment (michaelkain) [reply docker-compose/keycloak/Dockerfile]: 20.0.3 did not work for me, I tried, so I was happy to move to 20.0.0 and will wait for 21.0.0

File: docker-compose/keycloak/Dockerfile (hunk 3 above)
Comment (michaelkain) [reply docker-compose/keycloak/Dockerfile]: well done, that you removed it

**Comment by michaelkain** [comment]
as it works on qualif, tested by Anthony, and as I think it is a good idea to bring it live asap, I merge the PR. If the change to the "start"-flag in KC from "start-dev" causes a local problem in dev environment (see question by Anthony), we will fix afterwards. I agree with Anthony, better to start by default with production mode
